Iter5

WPGe opened the pull request.

WPGe committed:
first increment
- хендлер на серевера
- хранилище метрик
- сервис добавления метрик в хранилище

WPGe committed:
second increment
- изменил структуру проекта в internal, разбив его на слои:
handler, storage, repository, entity, agent
- покрыл тестами сервер и агент
- изменил метож хранения метрик в хранилище, теперь там хранится map структур Metric

WPGe committed:
3 increment
- внедрил библиотеку resty и chi
- для обработки Url использую chi
- для тестов вместо прямого вызова хендлера использую тестовый сервер

WPGe committed:
3 increment
- внедрил библиотеку resty и chi
- для обработки Url использую chi
- для тестов вместо прямого вызова хендлера использую тестовый сервер

WPGe committed:
4 increment
- добавлены флаги для сервера и агента

WPGe committed:
5 increment
- добавлены переменные окружения

vovean commented on cmd/agent/flag.go:
```diff
@@ -0,0 +1,40 @@
+		log.Fatal(err)
+	}
+
+	if cfg.Address != "" {
```
получается, переменные окружения имеют приоритет перед флагами? Может стоит использовать только один источник конфигурации? Если при запуске программы будут перемешиваться откуда какие параметры мы передаем, то пользователю будет легко запутаться.

Я бы предпочел переменные окружения, потому что 1) их использование соответствует 12 Factor Apps 2) переменные окружения можно прям как флаги задавать перед запуском каждой команды и они буду действовать только на этот запуск команды, а не на все окружение (принятый ответ вот здесь показывает, как это сделать: https://unix.stackexchange.com/questions/56444/how-do-i-set-an-environment-variable-on-the-command-line-and-have-it-appear-in-c)

WPGe replied:
флаги и переменные окружения используются совместно, т.к. это указано в задаче для инкрементов 4 и 5 и ожидается тестами
после прочтения статьи я все равно не вижу способа оставить только один вариант, притом, что названия для флагов и переменных окружения жестко указаны в задании и тестах

vovean replied:
согласен, из-за тестов придется так сделать

но хорошо, что статью прочитала)

vovean commented on cmd/agent/main.go:
```diff
@@ -1,3 +1,18 @@
+
+	stopCh := make(chan struct{})
+	go agent.MetricAgent(repo, "http://"+flagRunAddr+"/update", flagReportInterval, flagPollInterval, stopCh)
+	select {}
```
хмм, остаток от какой-то нереализованной идеи? Может стоит удалить

WPGe replied:
функция агента по условию задачи должна работать бесконечно, каждый flagPollInterval собирая метрики и каждый flagReportInterval отправляя метрики. select{} нужен, чтобы заблокировать горутину, чтобы она ждала выполнение описанных событий

WPGe replied:
убрал лишний вызов отдельной горутины, т.к. внутри функции уже есть бесконечный цикл

vovean commented on cmd/server/flag.go:
```diff
@@ -0,0 +1,17 @@
+	flag.StringVar(&flagRunAddr, "a", "localhost:8080", "address and port to run server")
+	flag.Parse()
+
+	if envRunAddr := os.Getenv("ADDRESS"); envRunAddr != "" {
```
в агенте используется структура конфига с автоматическим анмаршалингом значений из переменных окружения, а здесь ты работаешь напрямую с пакетом os. Как минимум из соображений консистентности подхода к написанию кода стоит использовать в обоих местах. К тому же подход со структурой более расширяемый

vovean commented on cmd/server/main.go:
```diff
@@ -1,3 +1,24 @@
+	r.Get("/value/{type}/{name}", handler.MetricGetHandler(repo))
+	r.Get("/", handler.MetricGetAllHandler(repo))
+
+	log.Fatal(http.ListenAndServe(flagRunAddr, r))
```
такое использование сервера не подразумевает graceful shutdown. Хорошей практикой является его использование. Подробнее про graceful shutdown и, в частности, о том, как его реализоваться с http.Server можно почитать здесь: https://www.rudderstack.com/blog/implementing-graceful-shutdown-in-go/

vovean commented on internal/agent/metric.go:
```diff
@@ -0,0 +1,103 @@
+	"time"
+)
+
+func MetricAgent(repo repository.MetricRepository, hookPath string, reportInterval int, pollInterval int, stopCh <-chan struct{}) {
```
может принимать time.Duration, а не int для параметров интервала? Потому что при использовании этой функции непонятно, в каких единицах измеряется интервал. И что, если я хочу интервал меньше секунды?

vovean commented on internal/agent/metric.go:
```diff
@@ -0,0 +1,103 @@
+func collectGaugeRuntimeMetrics() map[string]float64 {
+	var m runtime.MemStats
+	runtime.ReadMemStats(&m)
+	return map[string]float64{
```
каждый раз тут будет создаваться мапа, класться в кучу (разве что гошный компилятор произведет инлайнинг этой функции) и передаваться в вызывающую функцию. Потом сборщику мусора придется подчищать эту мапу.

Можно подумать в сторону того, чтобы использовать один экземпляр мапы и каждый раз заполнять его, а перед очередным считыванием метрик очищать его. Это, возможно, несколько усложнит программу, но улучшит работу с памятью

vovean replied:
Проверить, используется ли куча для передачи мапы (или происходит инлайнинг) можно включить вывод escape analysis в процессе компиляции: `go build -gcflags "-m -m"`

Подробнее:  https://habr.com/ru/articles/497994/

WPGe replied:
я решил в таком случае вообще избавиться от промежуточного хранения метрик в map и сразу записывать все в хранилище

vovean commented on internal/agent/metric.go:
```diff
@@ -0,0 +1,103 @@
+	}
+}
+
+func collectCounterRuntimeMetrics() map[string]int64 {
+	return map[string]int64{
+		"PollCount": 1,
+	}
+}
```
хм, странная логика...

vovean replied:
не уверен, что эта метрика делает, но использование захардкоженного значения - точно не расширяемое решение

WPGe replied:
тут метрика является счетчиком, который просто увеличивается на 1 при каждом сборе метрик.
Чтобы стало понятнее, не смог придумать ничего лучше, чем вынести это в отдельную функцию increasePollIteration, в названии которой явно указано, что в ней делается

vovean commented on internal/agent/metric.go:
```diff
@@ -0,0 +1,103 @@
+		req := resty.New().R()
+		req.Method = http.MethodPost
+		req.URL = url
+		_, err := req.Send()
```
было бы хорошо еще проверить response code и вывести его, если он не 2ХХ

vovean commented on internal/handler/update_handler.go:
```diff
@@ -0,0 +1,99 @@
+		metricValue := chi.URLParam(r, "value")
+
+		if metricType == "" || metricName == "" || metricValue == "" {
+			http.Error(w, "Некорректный формат URL", http.StatusNotFound)
```
может отдавать ошибки на английском языке? Как минимум на некоторых платформах могут возникнуть проблемы с отображением русского языка, как максимум, вдруг, этим проектом, однажды, будет пользоваться нерусскоговорящий пользователь

vovean commented on internal/handler/update_handler.go:
```diff
@@ -0,0 +1,99 @@
+
+		switch metricType {
+		case string(entity.Gauge):
+			if _, err := io.WriteString(w, fmt.Sprintf("%g", resultMetric.Value)); err != nil {
```
отличное применение пакета io!

vovean commented on internal/repository/metric.go:
```diff
@@ -0,0 +1,12 @@
+)
+
+type MetricRepository interface {
+	AddMetric(name string, metric entity.Metric)
```
не может ли быть тут ошибки? Например, в будущем, если будет использоваться какая-то БД, то она точно будет возвращать ошибку

vovean replied:
в методах ниже - аналогично

vovean commented on internal/storage/mem_storage.go:
```diff
@@ -0,0 +1,58 @@
+)
+
+type MemStorage struct {
+	mu      sync.Mutex
```
RWMutex? Можно делать RLock для операций чтения и Lock для записи

vovean commented on internal/storage/mem_storage.go:
```diff
@@ -0,0 +1,58 @@
+			existingMetric.Value = metric.Value
+		}
+		m.metrics[name] = existingMetric
+	} else {
```
можно сделать early return и обойтись без else, уменьшив уровень вложенности кода

WPGe committed:
review fix
- добавил использование конфига для сервера
- заменил тип параметров времени интервалов на time.Duration
- перевел текст ошибок на английский
- убрал лишнюю вложенность в функции добавления метрики
- переделал сбор метрики, чтобы лишний раз не создавалась map
- исправил работу метрики PollCount
- добавил проверку response code в методе отправки метрик

WPGe committed:
review fix
- добавлено возвращение ошибок в методах хранилища на будущее

WPGe committed:
review fix
- заменил метод сбора метрик, отказавший от промежуточного хранения в map
- Вынес увеличение метрики кол-ва прогонов сбора метрик в отдельный метод

WPGe committed:
review fix
- добавлена поддержка graceful shutdown для сервера

WPGe committed:
review fix
- убрал лишний вызов отдельной горутины, т.к. внутри функции уже есть бесконечный цикл
